backport: bitcoin#10244, bitcoin#18123, bitcoin#12906 + Parts of bitcoin#11403

xdustinface opened the pull request:
This is mainly the backport of bitcoin#10244 (replaces #3638) which is actually more than enough for one PR but i added bitcoin#18123 because it fixes a bug introduced in bitcoin#10244 and its just a simple diff.

I added the dash related changes into the related commits of the original PR where it made sense.. if i would do this PR again i would separate them probably lol.. but well, its done like this now..hope its not too hard to review 🙈 

352e7d3 unfortunately ended up being a bit too much imo.. besides removing the direct calls it also contains some refactoring.. should have done it in multiple steps but didn't expect it to become that big in the beginning.. i hope you can still get the reasonings for the changes.. if not, just ask :) 

~Based on #3680~

PastaPastaPasta commented:
Can you rebase/provide guidance on if this is ready for review?

xdustinface commented:
> Based on #3680

^^^ waiting for that to get your review before i rebase here.

xdustinface commented:
Ok i guess this one is ready now! Happy reviewing 🙈

UdjinM6 commented:
should revert/drop 9a07ca7 and backport bitcoin#12906 instead

xdustinface commented:
> should revert/drop [9a07ca7](https://github.com/dashpay/dash/commit/9a07ca797ef0013a84d62e24c547db1b37f579a0) and backport [bitcoin#12906](https://github.com/bitcoin/bitcoin/pull/12906) instead

Done 👍

UdjinM6 commented:
hmm.. would prefer it backported as a scripted-diff commit tbh

UdjinM6 commented on src/interfaces/node.cpp:
```diff
@@ -0,0 +1,449 @@
+    {
+        return MakeHandler(::uiInterface.NotifyBlockTip.connect([fn](bool initial_download, const CBlockIndex* block) {
+            fn(initial_download, block->nHeight, block->GetBlockTime(), block->GetBlockHash().ToString(),
+                GuessVerificationProgress(::Params().TxData(), block));
```
```suggestion
                GuessVerificationProgress(Params().TxData(), block));
```

xdustinface commented on src/wallet/wallet.cpp:
```diff
@@ -44,7 +44,7 @@
 
 static CCriticalSection cs_wallets;
-static std::vector<CWallet*> vpwallets GUARDED_BY(cs_wallets);
+std::vector<CWallet*> vpwallets GUARDED_BY(cs_wallets);
```
Good question! See 800029f2de473b7d874c910162e4215d2b2c0f49

xdustinface replied:
> Good question! See [800029f](https://github.com/dashpay/dash/commit/800029f2de473b7d874c910162e4215d2b2c0f49)

PastaPastaPasta replied:
it should probably be squashed into this commit then

xdustinface commented on src/qt/paymentserver.cpp:
```diff
@@ -621,29 +622,24 @@ void PaymentServer::fetchPaymentACK(CWallet* wallet, const SendCoinsRecipient& r
-        CScript s = GetScriptForDestination(*refundAddresses.begin());
+    CPubKey newKey;
+    if (walletModel->wallet().getKeyFromPool(false /* internal */, newKey)) {
+        // BIP70 requests encode the scriptPubKey directly, so we are not restricted to address
```
Hm yeah i think the reason i added this was just because it was a conflict there but this part looked more reasonable + it prevented the need to create an interface call for `GetAccountAddresses` + it leads to less conflicts in the future this way. So we can leave it as it is? How would you mention the partial backport in this case? Like in the code, PR title or what?

xdustinface replied:
See 36cd11f

PastaPastaPasta replied:
please change the name of the commit to start with "Partial Backport #XXXX"

xdustinface commented on src/interface/wallet.cpp:
```diff
@@ -51,6 +52,65 @@ class PendingWalletTxImpl : public PendingWalletTx
+    result.txin_is_mine.reserve(wtx.tx->vin.size());
+    for (const auto& txin : wtx.tx->vin) {
+        result.txin_is_mine.emplace_back(wallet.IsMine(txin));
+        if (result.txin_is_mine.back() && wallet.IsDenominated(txin.prevout)) {
```
See 05720a67d90b0c1a69f684791f2e32e575bb54eb

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <qt/clientmodel.h>
 #include <qt/guiutil.h>
 #include <init.h>
+#include <interface/node.h>
```
upstream didn't add this

xdustinface replied:
See 395626d

PastaPastaPasta replied:
squash as well

PastaPastaPasta commented on src/qt/overviewpage.h:
```diff
@@ -50,16 +49,10 @@ public Q_SLOTS:
     int nDisplayUnit;
     bool fShowAdvancedPSUI;
     int cachedNumISLocks;
+    interface::WalletBalances m_balances;
```
this should be under `WalletModel *walletModel;` imo

xdustinface replied:
YOU want to break up with upstream? :D

PastaPastaPasta replied:
No, that would be most consistent with upstream imo

xdustinface replied:
ahh got you wrong initially :D not sure why but i thought you want to move the `m_balances` stuff into `WalletModel`

PastaPastaPasta commented on src/interface/wallet.cpp:
```diff
@@ -127,6 +127,10 @@ class WalletImpl : public Wallet
 
+    void markDirty() override
+    {
+        m_wallet.MarkDirty();
```
what exactly does MarkDirty do?

xdustinface replied:
Same as before, marking the wallet dirty :D Basically just forcing to recalculate cached balances of the wallet/wallet transactions.

PastaPastaPasta commented on src/privatesend/privatesend-client.h:
```diff
@@ -7,8 +7,6 @@
 
 #include <privatesend/privatesend-util.h>
 #include <privatesend/privatesend.h>
-#include <wallet/wallet.h>
```
e3f0108d7bae13969d90094e5f85ee6a02b0f607 message probably shouldn't have prefix exclusively be `qt`

xdustinface replied:
See e031596

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1434,21 +1435,28 @@ bool CPrivateSendClientSession::MakeCollateralAmounts(const CompactTallyItem& ta
 
     // Skip way too tiny amounts. Smallest we want is minimum collateral amount in a one output tx
-    if (!txBuilder.CouldAddOutput(CPrivateSend::GetCollateralAmount())) {
+    if (!txBuilder.CouldAddOutput(CPrivateSend::GetMinCollateralAmount())) {
```
Why is GetCollateralAmount getting refactored in this commit?

xdustinface replied:
Hmm good question, guess i just thought it makes sense to give it better name in the same run as we introduce it in `interfaces`. Concerns?

PastaPastaPasta replied:
Yeah, should minimize the diff on that imo. Please revert. I'm not against it in another PR though

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -321,12 +321,10 @@ void OverviewPage::showOutOfSyncWarning(bool fShow)
-    if(!clientModel->masternode().isBlockchainSynced() || ShutdownRequested()) return;
-
-    if(!walletModel) return;
+    if(ShutdownRequested() || !walletModel || !clientModel || !clientModel->masternode().isBlockchainSynced()) return;
```
Since you're already changing this and it's dash specific
```suggestion
    if (ShutdownRequested() || !walletModel || !clientModel || !clientModel->masternode().isBlockchainSynced()) return;
```

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -421,22 +419,25 @@ void OverviewPage::updatePrivateSendProgress()
-void OverviewPage::updateAdvancedPSUI(bool fShowAdvancedPSUI) {
+void OverviewPage::updateAdvancedPSUI(bool fShowAdvancedPSUI)
+{
+    if(!walletModel || !clientModel || !clientModel->privateSendOptions().isEnabled()) return;
```
```suggestion
    if (!walletModel || !clientModel || !clientModel->privateSendOptions().isEnabled()) return;
```

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -421,22 +419,25 @@ void OverviewPage::updatePrivateSendProgress()
 void OverviewPage::privateSendStatus(bool fForce)
 {
-    if (!fForce && (!clientModel->masternode().isBlockchainSynced() || ShutdownRequested())) return;
+    if(!walletModel || !clientModel) return;
```
```suggestion
    if (!walletModel || !clientModel) return;
```

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -496,15 +497,8 @@ void OverviewPage::privateSendStatus(bool fForce)
-
     // We are processing more than 1 block per second, we'll just leave
-    if(nBestHeight > privateSendClientManager->nCachedNumBlocks && GetTime() - nLastDSProgressBlockTime <= 1) return;
+    if(nBestHeight > walletModel->privateSend().getCachedBlocks() && GetTime() - nLastDSProgressBlockTime <= 1) return;
```
```suggestion
    if (nBestHeight > walletModel->privateSend().getCachedBlocks() && GetTime() - nLastDSProgressBlockTime <= 1) return;
```

PastaPastaPasta commented on src/qt/overviewpage.cpp:
```diff
@@ -620,15 +614,9 @@ void OverviewPage::togglePrivateSend(){
-        const CAmount nMinAmount = CPrivateSend::GetSmallestDenomination() + CPrivateSend::GetMaxCollateralAmount();
+    if (!walletModel->privateSend().isMixing()) {
+        auto& options = walletModel->node().privateSendOptions();
+        const CAmount nMinAmount = options.getSmallestDenomination() + options.getMaxCollateralAmount();
```
Why max collateral amount? @UdjinM6 (I see you're maintaining old logic, so this isn't really a problem for you @xdustinface)

PastaPastaPasta commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2937,7 +2937,6 @@ UniValue loadwallet(const JSONRPCRequest& request)
     if (!wallet) {
         throw JSONRPCError(RPC_WALLET_ERROR, "Wallet loading failed.");
     }
-    AddWallet(wallet);
```
This feels VERY unrelated

xdustinface replied:
Part of https://github.com/dashpay/dash/pull/3682#discussion_r498753310, Removed it there because `AddWallet` is called already in `CreateWalletFromFile` with the related changes

PastaPastaPasta replied:
Are these changes a necessity to be in this PR? they feel like they should be broken out

xdustinface replied:
It will simply not work without those changes in here unless we refactor `interfaces::PrivateSend::Client` and/or move it out of `interfaces::Wallet`. But IMO its pretty nice the way they are cooperated right now and i would definitely vote to leave it as it is but im also always open for concrete suggestions.

PastaPastaPasta commented on src/wallet/wallet.cpp:
```diff
@@ -5120,12 +5118,23 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
     auto temp_wallet = MakeUnique<CWallet>(name, WalletDatabase::Create(path));
     CWallet *walletInstance = temp_wallet.get();
-    DBErrors nLoadWalletRet = walletInstance->LoadWallet(fFirstRun);
+    AddWallet(walletInstance);
```
and below

xdustinface replied:
Those are to make sure there is always consistency between `vpwallets` and `privateSendClientManagers`. This is required because `interfaces::CPrivateSend::Client` is member of `interfaces::Wallet` and therefor requires to have the wallet already in `privateSendClientManagers` before the first `interfaces::Wallet` instance gets created which happens during wallet loading when `uiInterface.LoadWallet(walletInstance)` is called.

PastaPastaPasta commented on src/interfaces/node.cpp:
```diff
@@ -359,7 +359,7 @@ class NodeImpl : public Node
 #ifdef ENABLE_WALLET
         std::vector<std::unique_ptr<Wallet>> wallets;
-        for (CWalletRef wallet : ::vpwallets) {
+        for (CWalletRef wallet : GetWallets()) {
```
can you explain 800029f2de473b7d874c910162e4215d2b2c0f49?

xdustinface replied:
Just reverting some accidental changes

PastaPastaPasta commented on src/interfaces/wallet.cpp:
```diff
@@ -87,7 +89,7 @@ WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
     result.value_map = wtx.mapValue;
     result.is_coinbase = wtx.IsCoinBase();
-    result.is_denominate = fIsDenomInFromMe && fAllDenomOut && fIsDenomOutFromMe;
+    result.is_denominate = fIsDenomInFromMe && fAllDenomOut && fAllDenom;
```
fAllDenomOut is undefined

xdustinface replied:
See f67ea7d

PastaPastaPasta commented on src/qt/addresstablemodel.cpp:
```diff
@@ -8,7 +8,7 @@
 #include <qt/walletmodel.h>
 
-#include <interfaces/node.h>
+#include <interfaces/wallet.h>
```
why?

xdustinface replied:
Not sure where this came from but does not longer exist this way

PastaPastaPasta commented on src/qt/rpcconsole.cpp:
```diff
@@ -1343,7 +1343,7 @@ void RPCConsole::banSelectedNode(int bantime)
 	const CNodeCombinedStats *stats = clientModel->getPeerTableModel()->getNodeStats(detailNodeRow);
 	if(stats) {
-            m_node.ban(stats->nodeStats.addr, BanReasonManuallyAdded, bantime);
+        m_node.ban(stats->nodeStats.addr, BanReasonManuallyAdded, bantime);
```
we should keep upstream's extra spaces, but even if we did decide to drop them, this only has two spaces not four

xdustinface commented on src/qt/rpcconsole.cpp:
```diff
@@ -908,7 +910,7 @@ void RPCConsole::updateNetworkState()
     connections += tr("In:") + " " + QString::number(clientModel->getNumConnections(CONNECTIONS_IN)) + " / ";
     connections += tr("Out:") + " " + QString::number(clientModel->getNumConnections(CONNECTIONS_OUT)) + ")";
 
-    if(!clientModel->getNetworkActive()) {
```
see #3720

xdustinface commented on src/interface/node.cpp:
```diff
@@ -121,7 +121,7 @@ class NodeImpl : public Node
+        return GuessVerificationProgress(Params().TxData(), tip);
     }
     bool isInitialBlockDownload() override { return IsInitialBlockDownload(); }
     bool getReindex() override { return ::fReindex; }
```
Can you link here what you mean?

xdustinface commented on src/qt/intro.cpp:
```diff
@@ -239,7 +240,7 @@ bool Intro::pickDataDirectory()
      */
     if(dataDir != dataDirDefaultCurrent)
-        gArgs.SoftSetArg("-datadir", GUIUtil::qstringToBoostPath(dataDir).string()); // use OS locale for path setting
+        node.softSetArg("-datadir", GUIUtil::qstringToBoostPath(dataDir).string()); // use OS locale for path setting
```
See 668b67bd83f634f7a59cc094bb7fc5967ffb9872

PastaPastaPasta replied:
please squash into this commit

xdustinface commented on src/qt/walletmodeltransaction.cpp:
```diff
@@ -55,7 +54,7 @@ void WalletModelTransaction::reassignAmounts()
                 const unsigned char* scriptStr = (const unsigned char*)out.script().data();
                 CScript scriptPubKey(scriptStr, scriptStr+out.script().size());
-                for (const auto& txout : walletTransaction->vout) {
+                for (const auto& txout : wtx->get().vout) {
```
Matching upstream like how? All surrounding code isn't upsteam it seems?

xdustinface commented on src/qt/splashscreen.cpp:
```diff
@@ -177,36 +177,35 @@ static void ShowProgress(SplashScreen *splash, const std::string &title, int nPr
-        pwallet->ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, false));
+    m_handler_init_message->disconnect();
+    m_handler_show_progress->disconnect();
+    m_handler_load_wallet->disconnect();
```
Here https://github.com/dashpay/dash/pull/3462#issuecomment-620510626

xdustinface commented on src/qt/walletmodeltransaction.cpp:
```diff
@@ -18,14 +17,14 @@ QList<SendCoinsRecipient> WalletModelTransaction::getRecipients() const
 unsigned int WalletModelTransaction::getTransactionSize()
 {
-    return (!walletTransaction ? 0 : (::GetSerializeSize(*walletTransaction, SER_NETWORK, PROTOCOL_VERSION)));
+    return (!wtx ? 0 : (::GetSerializeSize(wtx->get(), SER_NETWORK, PROTOCOL_VERSION)));
```
See f4101b975f23ef146cacd26a5a3ca775e7392539

PastaPastaPasta replied:
Hmm, not really, I was specifically making that change to match upstream. adding the explicit check breaks that consistency

xdustinface replied:
This line is anyway different in the upstream and will lead to conflicts, so why not follow rules we usually follow?

xdustinface commented on src/qt/coincontroldialog.cpp:
```diff
@@ -15,6 +15,7 @@
 
 #include <wallet/coincontrol.h>
 #include <init.h>
```
Can't remember if there was a reason :D Probably not, see b66ee134e425e22f8c400d2b7753dfa1233e1d87

PastaPastaPasta replied:
Should squash into this commit

xdustinface commented on src/qt/addresstablemodel.cpp:
```diff
@@ -8,7 +8,7 @@
 #include <qt/walletmodel.h>
 
-#include <wallet/wallet.h>
+#include <interfaces/node.h>
```
Reason for this?

xdustinface commented on src/qt/overviewpage.cpp:
```diff
@@ -704,10 +686,7 @@ void OverviewPage::DisablePrivateSendCompletely() {
-        return;
+    for (auto& wallet : clientModel->node().getWallets()) {
+        wallet->privateSend().stopMixing();
     }
```
See 16de64b7a0acc2843d0c800c17b6ba5458e097c3, applied the same logic to the `disableAutobackups` above

xdustinface commented on src/qt/walletmodel.cpp:
```diff
@@ -6,26 +6,20 @@
 
 #include <base58.h>
-#include <chain.h>
+
```
See 395626d

PastaPastaPasta replied:
I'd say squash that change into this commit

xdustinface commented on src/qt/addressbookpage.cpp:
```diff
@@ -241,9 +241,7 @@ void AddressBookPage::on_showAddressQRCode_clicked()
 
```
Hmm long time ago 🙈 but looks like its just because `QDialog` doesn't need it for anything, `OptionsModel`'s constructor changed, so i just dropped the related stuff it seems. Also a7024cf28cb910c17d71968c9b816286b6fa9f00 now.

xdustinface commented on src/qt/walletmodel.h:
```diff
@@ -203,42 +189,39 @@ class WalletModel : public QObject
     int getNumISLocks() const;
 
     int getRealOutpointPrivateSendRounds(const COutPoint& outpoint) const;
+
```
See 2a1abe4f7ca656e9925e5999ae626f49ca1a0a03

PastaPastaPasta replied:
go ahead and squash it into this commit

xdustinface commented on src/interfaces/node.h:
```diff
@@ -0,0 +1,338 @@
+};
+
+//! Interface for the src/masternode part of a dash node (dashd process).
+class Masternode
```
I totally agree, this "masternode sync" related stuff needs to be refactored 😄 Thoughts about just moving 

```
    virtual bool isBlockchainSynced() = 0;
    virtual bool isSynced() = 0;
    virtual std::string getSyncStatus() =  0;
```

into `interfaces::Node` ? Like having calls like `node.isSynced()` 

Kind of breaks with my initial idea i.e. having an interface for each subdirectory stuff but it somehow sounds more reasonable to have those methods in the `Node` interface..

xdustinface replied:
If there are any concerns about moving it into `Node` i would probably just do the `Masternode::Sync` thing for now and revisit this in a future PR.

xdustinface replied:
@UdjinM6 ^

UdjinM6 replied:
I'd say `Masternode::Sync` is the way to go.

UdjinM6 commented on src/interfaces/wallet.cpp:
```diff
@@ -65,7 +65,7 @@ WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
         result.txin_is_mine.emplace_back(wallet.IsMine(txin));
         // NOTE: We check only inputs our wallet knows about, cause IsDenominated returns "false" for non-wallet txes
-        // and finding/fetching the actual tx from blocks db locks cs_main which we should (and can) avoid here.
+        // and finding/fetching the actual tx from mapWallet locks cs_wallet which we should (and can) avoid here.
```
No, I mean finding that tx requires scanning blocks db via GetTransaction which locks cs_main

xdustinface replied:
Hm i guess i can't really follow your thoughts here right now, can you explain how/where exactly 004752b avoids calling `GetTransaction`?

UdjinM6 replied:
I mean `wallet.IsDenominated(...)` works only for inputs that are known to _our_ wallet and many inputs in mixing tx do not belong to it. To check all inputs we can do f4e4973aef (and then aaea64b170 some parts of which are required anyway ;) ). But I just realized that `cs_main` is locked whenever we call `MakeWalletTx` and this lock inside `GetTransaction` is not an issue really, accessing tx index is probably fine too 🤔

xdustinface replied:
@UdjinM6 Can you check ac3a8a0? I thiiiink that should be good enough?

xdustinface commented on src/interfaces/wallet.cpp:
```diff
@@ -0,0 +1,525 @@
+    result.time = wtx.GetTxTime();
+    result.value_map = wtx.mapValue;
+    result.is_coinbase = wtx.IsCoinBase();
+    result.is_denominate = fIsDenomInFromMe && fIsDenomOutFromMe && fAllDenom;
```
@UdjinM6 Picked! Thoughts about ecf497a on top? And i might have misunderstood something there but i think the comment should rather be like in 69356b5?

UdjinM6 replied:
Looks good besides 69356b5, see my comment below

UdjinM6 commented on src/qt/overviewpage.cpp:
```diff
@@ -289,6 +273,13 @@ void OverviewPage::setWalletModel(WalletModel *model)
+        // we'll handle automatic backups and user warnings in privateSendStatus()
+        walletModel->privateSend().disableAutobackups();
+
+        timer = new QTimer(this);
+        connect(timer, SIGNAL(timeout()), this, SLOT(privateSendStatus()));
```
This breaks UI updates during mixing for me. Moving these 2 lines back to `OverviewPage` 
 ctor fixes the issue.

xdustinface replied:
Fixed and squashed already.. wonder why i even changed this :D

UdjinM6 commented on src/interfaces/wallet.cpp:
```diff
@@ -79,15 +75,19 @@ WalletTx MakeWalletTx(CWallet& wallet, const CWalletTx& wtx)
     result.time = wtx.GetTxTime();
     result.value_map = wtx.mapValue;
     result.is_coinbase = wtx.IsCoinBase();
+    result.is_denominate = wtx.tx->vin.size() == wtx.tx->vout.size() && // Number of inputs is same as number of outputs
+                           (result.credit - result.debit) == 0 && // Transaction pays no tx fee
```
nit: imo this is a bit more readable
```suggestion
    result.is_denominate = (wtx.tx->vin.size() == wtx.tx->vout.size()) && // Number of inputs is same as number of outputs
                           (result.credit - result.debit == 0) && // Transaction pays no tx fee
```

UdjinM6 reviewed (commented):
0fb1a05 works for me :) Maybe add a comment that we only do some simple checks here which should be enough to reliably detect mixing txes for this specific wallet and that this relies on PS subsystem verifying the tx earlier (during mixing) and wallet not participating in txes that won't fit more strict rules or smth like that.
+ see one tiny nit below

UdjinM6 commented:
oh, and you still need smth like b22ad872f7

xdustinface commented:
@PastaPastaPasta Squashed all stuff, renamed that backport commit one more time, renamed `CPrivateSend::GetMinCollateralAmount` to `CPrivateSend::GetCollateralAmount`.

@UdjinM6 Picked 0fb1a05, added the comment about the simplified denom checks, squashed all denom related commits,  squashed the PS timer issue, and i added d624fbf0dafc954e8444d120451e56ffe441f142, 0abdc351d5a3eb144ea06e6e84f9c6db0802527f, 76bea286d172ab8f4ffa1333d396bcfb4f7a9a9a as replacement for https://github.com/dashpay/dash/pull/3682#discussion_r497610846 thoughts?

UdjinM6 commented:
Hmm... I like d624fbf (the idea of caching it in `CMerkleTx`) but not 0abdc35 and 76bea28 tbh. Why don't we simply implement (a simplified version of) the original logic in `CMerkleTx::IsLockedByInstantSend()` instead? i.e.
```
        if (status.lockedByChainLocks) {
            if (oldLockedByChainLocks != status.lockedByChainLocks) {
                status.lockedByInstantSend = wtx.IsLockedByInstantSend();
            } else {
                status.lockedByInstantSend = false;
            }
        } else if (!status.lockedByInstantSend) {
            status.lockedByInstantSend = wtx.IsLockedByInstantSend();
        }
```
->
```
bool CMerkleTx::IsLockedByInstantSend() const
{
    if (fIsChainlocked) {
        fIsInstantSendLocked = false;
    } else if (!fIsInstantSendLocked) {	
        fIsInstantSendLocked = llmq::quorumInstantSendManager->IsLocked(GetHash());
    }
    return fIsInstantSendLocked;
}
```
(`if (oldLockedByChainLocks != status.lockedByChainLocks) {` part wasn't really useful anyway I think)

xdustinface commented:
> Hmm... I like [d624fbf](https://github.com/dashpay/dash/commit/d624fbf0dafc954e8444d120451e56ffe441f142) (the idea of caching it in `CMerkleTx`) but not [0abdc35](https://github.com/dashpay/dash/commit/0abdc351d5a3eb144ea06e6e84f9c6db0802527f) and [76bea28](https://github.com/dashpay/dash/commit/76bea286d172ab8f4ffa1333d396bcfb4f7a9a9a) tbh. Why don't we simply implement (a simplified version of) the original logic in `CMerkleTx::IsLockedByInstantSend()` instead? i.e.
 
The idea was to allow to use the cached value also in other places where `CMerkleTx::IsLockedByInstantSend` is used and called frequently. But since i didn't investigate a lot if that would even make sense at the end, lets go with your solution, looks good :) 

> 
> (`if (oldLockedByChainLocks != status.lockedByChainLocks) {` part wasn't really useful anyway I think)

Agreed :D 

The coming force push includes the `interfaces::Masternode` refactoring, drops  0abdc35 and 76bea28, adds the suggested changes above, squashes the remaining stuff

UdjinM6 commented:
Looks good 👍  Will test it a bit.

UdjinM6 commented:
Found an issue with my earlier suggestion you implemented in b1bc1f1 - `numBlocks` was fetched from `tryGetTxStatus`, so swapping the order functions are called in resulted in `numBlocks` always being 0 (undefined) for `statusUpdateNeeded` which means "yes" regardless of any other conditions. These issues (+ using hash instead of height but let's ignore it for now) are fixed in a series of btc PRs much later and can't be backported that easy imo. The simplest fix I can think of is 4f517e787e. It _should not_ cause any serious merge conflicts in the future as far as I can tell.

xdustinface commented:
> The simplest fix I can think of is [4f517e7](https://github.com/dashpay/dash/commit/4f517e787eafa99b2a4ceb02a56dddb7046fe3b0). It _should not_ cause any serious merge conflicts in the future as far as I can tell.

Looks good, but one question.. the extra `pollBalanceChanged` call in `WalletModel` is like trying to initialize cached blocks probably?

UdjinM6 commented:
Good question :) Yes, otherwise `updateStatus` is executed twice - first time with some initial value (0) and then with the actual number of blocks 250ms later. But the reason why this happens is that there is an inconsistency in initial values between transaction record status and caches values in transaction and wallet models. I _think_ 3433ec6caf should fix it.

xdustinface commented:
> I _think_ [3433ec6](https://github.com/dashpay/dash/commit/3433ec6cafde640de3de7d9fa84433e0b28c1c07) should fix it.

That looks better to me jup :)

UdjinM6 reviewed (dismissed):
Everything seems to be working as expected now.

ACK

PastaPastaPasta commented:
overall ACK, however I would like to see some commits merged. IMO all commits directly from 10244 should be one commit, then dash specific 10244 as one commit, then 18123 (then any associated review commits), then 12906 (then any associated review commits), then 11403 (then any associated review commits). It's difficult for me to tell at this point which review commits correspond to which actual backports

UdjinM6 commented:
I don't think squashing multiple focused 10244 commits into one huge commit is a good idea, would rather prefer them as is. Squashing performance fixes after 18123 (b1bc1f1 + b46e65b + 34a146c) makes sense though. Squashing aa2ebe1 + 60a8584 might also be a good idea. 11403 and 12906 need no additional work imo.

xdustinface commented:
> I don't think squashing multiple focused 10244 commits into one huge commit is a good idea would rather prefer them as is.

Agreed, right now i honestly don't see any benefit by doing this @PastaPastaPasta  

> Squashing performance fixes after 18123 ([b1bc1f1](https://github.com/dashpay/dash/commit/b1bc1f164701c529b584af346b75031b86da4057) + [b46e65b](https://github.com/dashpay/dash/commit/b46e65b3c1a80899a817c6393576e2c4dd1dc6d0) + [34a146c](https://github.com/dashpay/dash/commit/34a146c5c8e5e5242c92b297f70ac9fbe82fc842)) makes sense though. 

Done in 5020a68

> Squashing [aa2ebe1](https://github.com/dashpay/dash/commit/aa2ebe19786f4426239755bca5afd513fac989ec) + [60a8584](https://github.com/dashpay/dash/commit/60a858471694e2f6dfefbefaeff97b35deb66dbe) might also be a good idea. 

Done

UdjinM6 reviewed (approved):
utACK

PastaPastaPasta reviewed (approved):
ACK
